plutus-core/plutus-ir/src/PlutusIR/Transform/Rewrite.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+module PlutusIR.Transform.Rewrite where
+
+import PlutusCore.Default
+import PlutusIR.Analysis.Builtins
+import PlutusIR
+import Data.Monoid
+
+import Control.Lens
+
+userRewrite :: (Semigroup a, t ~ Term tyname name uni fun a)
+            => BuiltinsInfo uni fun
+            -> t
+            -> t
+userRewrite bi t =
+    let RewriteRules f = bi^.rewriteRules
+    in transformOf termSubterms (appEndo $ getDual f) t
+
+defaultUniRewriteRules :: RewriteRules DefaultUni DefaultFun
+defaultUniRewriteRules = RewriteRules $ combineRules
+    -- rules are applied from left to right because of Dual
+    [ decodeEncode
```

Comment:
As discussed, I'm not 100% sure this is safe. Maybe we should do `head (mkCons x xs) = let !_ = xs in x` or something?

Reply:
Y I am also not certain.

Reply:
Turns out that this rewrite is safe in Haskell because  Text.fromString will handle gracefully non-unicode literal characters, see: https://hackage.haskell.org/package/text-2.1/docs/Data-Text.html#g:2

It is not safe currently in our case because:
1) unicode text is broken on plutus-tx
2) we don't do this graceful transformation in ghc-core that Haskell normally does. We have to replicate this behaviour in our plugin: https://hackage.haskell.org/package/text-2.1/docs/src/Data.Text.Internal.html#pack